### PR TITLE
POC: Replace asyncAction with async rule

### PR DESCRIPTION
**Pull request by lchanmann**
## Ref
- https://nanopay.atlassian.net/browse/NP-5796

## Changes
- Replace asyncAction with async flag
- Async rules are always run after put/remove 
- Async rules are run sequentially in a new thread
- Change to `obj` in async rules actions will not be saved the object to DAO, would require putting to the DAO explicitly to save the object

### CODE REVIEW

File: src/foam/nanos/dig/DUGRule.js [hunk 1]
```diff
@@ -183,14 +183,11 @@ foam.CLASS({
       section: 'basicInfo'
     },
     {
-      name: 'action',
-      documentation: `Action is not used for DUGRules, see asyncAction instead`,
-      hidden: true,
-      section: 'basicInfo',
-      networkTransient: true
+      name: 'async',
```
Comment (Nauna) [inline src/foam/nanos/dig/DUGRule.js:186]: adding class: 'Boolean' makes this more clear

File: src/foam/nanos/ruler/RulerProbe.js [hunk 2]
```diff
@@ -36,6 +36,16 @@ foam.CLASS({
       of: 'foam.nanos.dao.Operation',
       name: 'operation'
     },
+    {
+      class: 'Boolean',
+      name: 'after',
+      value: false,
```
Comment (Nauna) [inline src/foam/nanos/ruler/RulerProbe.js]: hmmm just looks odd since Boolean default is false

**Review by Nauna — COMMENTED** [review]
LGTM 👍

File: src/foam/nanos/ruler/RulerDAO.js [hunk 3]
```diff
@@ -49,6 +45,7 @@ foam.CLASS({
       of: 'foam.mlang.predicate.Predicate',
       name: 'createBefore',
       javaFactory: `return AND(
+  EQ(Rule.ASYNC, false),
```
Comment (tharmaman) [inline src/foam/nanos/ruler/RulerDAO.js:48]: More of a pet peeve but the indenting in the javaFactories for this model look off

File: src/foam/nanos/ruler/RulerDAO.js [hunk 4]
```diff
@@ -106,8 +107,42 @@ foam.CLASS({
       of: 'foam.mlang.predicate.Predicate',
       name: 'removeAfter',
       javaFactory: `return AND(
+  EQ(Rule.ASYNC, false),
   EQ(Rule.OPERATION, Operation.REMOVE),
   EQ(Rule.AFTER, true)
+);`
```
Comment (tharmaman) [inline src/foam/nanos/ruler/RulerDAO.js:113]: again here think the body of javaFactory just needs an indent

File: src/foam/nanos/ruler/RulerProbe.js [hunk 5]
```diff
@@ -36,6 +36,16 @@ foam.CLASS({
...
+    },
+    {
+      class: 'Boolean',
+      name: 'async',
+      value: false
```
Comment (tharmaman) [inline src/foam/nanos/ruler/RulerProbe.js]: Yeah agree with anna for the default value for boolean

File: src/foam/nanos/ruler/RuleEngine.java [hunk 6]
```diff
@@ -165,32 +159,56 @@ public void stop() {
...
+      // the eligibility to execute the rule action.
+      //
+      // NOTE: There is no need to reload the "rule" object and re-check
+      // isActive and permission as it was the rule at the time RuleEngine.execute()
+      // is called that the rule engine honors and should commit to, not the
+      // future version of the same rule.
+      var nu = reloadObject(obj);
+      if ( ! rule.f(x, nu, oldObj) ) return;
+
+      PM pm = PM.create(getX(), RulerDAO.getOwnClassInfo(), "ASYNC: " + rule.getDaoKey() + ": " + rule.getId());
```
Comment (jlhughes) [inline src/foam/nanos/ruler/RuleEngine.java]:
```suggestion
      PM pm = PM.create(getX(), RulerDAO.getOwnClassInfo(), "ASYNC", rule.getDaoKey(), rule.getId());
```

File: src/foam/nanos/ruler/RulerDAO.js (hunk 3 above)
Comment (lchanmann) [reply src/foam/nanos/ruler/RulerDAO.js:48]: The indentation is for the generated java code.

File: src/foam/nanos/ruler/RulerDAO.js (hunk 4 above)
Comment (lchanmann) [reply src/foam/nanos/ruler/RulerDAO.js:113]: same as above, indentation for the generated java code.

File: src/foam/nanos/ruler/RuleEngine.java [hunk 7]
```diff
@@ -65,12 +68,11 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
       PM pm = PM.create(getX(), RulerDAO.getOwnClassInfo(), rule.getDaoKey() + ": " + rule.getId());
```
Comment (jlhughes) [inline src/foam/nanos/ruler/RuleEngine.java]:
```suggestion
      PM pm = PM.create(getX(), RulerDAO.getOwnClassInfo(), rule.getDaoKey(), rule.getId());
```

**Review by jlhughes — COMMENTED** [review]
LGTM

File: src/foam/nanos/ruler/Rule.js [hunk 8]
```diff
@@ -140,6 +140,14 @@
       documentation: 'Defines if the rule needs to be applied before or after operation is completed'+
       'E.g. on dao.put: before object was stored in a dao or after.'
     },
+    {
+      class: 'Boolean',
+      name: 'async',
```
Comment (kgrgreer) [inline src/foam/nanos/ruler/Rule.js:145]: Does this cause any issues with the async reserved word in JS? Can you configure this through the GUI? I just did an experiment and it seems fine.

**Pull request merged**